### PR TITLE
Unify More Loot!

### LINKED ISSUE #323: Vanilla Raw Ores Found As Loot
### Modpack Version

1.6.0

### Describe your issue.

When looting an Integrated Dungeons castle I found vanilla raw copper ore, raw iron ore, and raw gold ore in some of the Lootr chests.

### Crash Report

_No response_

### Latest Log

_No response_

### Have you modified the modpack?

Yes

### User Modifications

Added InventoryHUD+, this is the only change I've done.

### Did the issue happen in singleplayer or on a server?

Singleplayer

### Discord Username

_No response_

### PR DESCRIPTION
**Pull request by Kanzaji**
*Should* resolve #323
The script has been reworked for ease of use, less amount of code and more flexibility without much work to add stuff in the future!

From my testing it appears to work fine, but future testing for edge-cases is required >.>
Most important to test more before release:

- Crushed ores in loot chests
- Gems from entities.

### CODE REVIEW

File: kubejs/server_scripts/base/unification/unify_loot.js [hunk 1]
```diff
@@ -1,196 +1,80 @@
...
-                    }
-                }
-            }
-            console.error(
-                'LootJS Unification Loot Modifier triggered for Ingots, but none of the required tags were found in the item: ' +
-                    ItemStack.getId()
-            );
-            return ItemStack;
-        })
+        .modifyLoot(Ingredient.of('#forge:ingots'), (ItemStack) => {return unifyLoot("forge:ingots", "ingot", ItemStack, null, true)})
```
Comment (NielsPilgaard) [inline kubejs/server_scripts/base/unification/unify_loot.js:42]: please use camelcase for variables :)

File: kubejs/server_scripts/base/unification/unify_loot.js [hunk 2]
```diff
@@ -1,196 +1,80 @@
...
-    // Replace ingots / nuggets / Storage Blocks / Gems in loot chests!
+    // How to use unifyLoot()
+    // tag              -> String - Tag to search for (accepts both tags with and without #)
+    // suffix           -> String - suffix for the EE variant of the item
+    // prefix           -> String - prefix for the EE variant of the item (for example, when you want to unify crushed ores (...:crushed_iron_ore))
+    // ItemStack        -> ItemStack - Original ItemStack that was meant to be dropped
+    // replaceWithRaw   -> Boolean - Should it replace the drop with Raw ore, or return unified version of the item?
+    // count            -> integer - if replaceWithRaw is True, then specifies the amount of raw ore given for each item.
+    function unifyLoot(tag, suffix, ItemStack, prefix, replaceWithRaw, count) {
+        prefix = (typeof prefix == undefined || prefix == null)? "": prefix + "_";
```
Comment (NielsPilgaard) [inline kubejs/server_scripts/base/unification/unify_loot.js:11]: I think you can just say `!prefix` instead of `(typeof prefix == undefined || prefix == null)`

File: kubejs/server_scripts/base/unification/unify_loot.js [hunk 3]
```diff
@@ -1,196 +1,80 @@
...
+        if (ItemStack.getId().startsWith('emendatusenigmatica:') && !replaceWithRaw) {return ItemStack;}
+        tag = (tag.startsWith("#")? tag.substring(1): tag) + "/"
+        let iterator = ItemStack.getTags().iterator();
+        while (iterator.hasNext()) {
+            let tagString = iterator.next().location().toString();
+            if (tagString.startsWith(tag)) {
+                let tagSubString = tagString.substring(tag.lastIndexOf("/")+1);
+                if (Item.exists(`emendatusenigmatica:raw_${tagSubString}`) && replaceWithRaw) {
+                    return Item.of(`emendatusenigmatica:raw_${tagSubString}`, count);
+                } else if (Item.exists(`emendatusenigmatica:${tagSubString}_${suffix}`)) {
```
Comment (NielsPilgaard) [inline kubejs/server_scripts/base/unification/unify_loot.js:24]: No need for `else` when there's a return in the if above

File: kubejs/server_scripts/base/unification/unify_loot.js [hunk 4]
```diff
@@ -1,196 +1,80 @@
...
+                    return Item.of(`emendatusenigmatica:${prefix}${tagSubString}_${suffix}`, ItemStack.getCount());
+                } else {
```
Comment (NielsPilgaard) [inline kubejs/server_scripts/base/unification/unify_loot.js:26]: No need for this else, since there's a return above

File: kubejs/server_scripts/base/unification/unify_loot.js [hunk 5]
```diff
@@ -1,196 +1,80 @@
...
-                    }
-                }
-            }
-            console.error(
-                'LootJS Unification Loot Modifier triggered for Gems, but none of the required tags were found in the item: ' +
-                    ItemStack.getId()
-            );
-            return ItemStack;
-        });
+        .modifyLoot(Ingredient.of('#forge:storage_blocks'), (ItemStack) => {return unifyLoot("forge:storage_blocks", "block", ItemStack, null, true, ItemStack.getCount() * (Math.floor(Math.random() * 3) + 3))})
```
Comment (NielsPilgaard) [inline kubejs/server_scripts/base/unification/unify_loot.js:46]: If you omit `{ }` the return can be removed in lambda expressions

File: kubejs/server_scripts/base/unification/unify_loot.js (hunk 1 above)
Comment (Kanzaji) [reply kubejs/server_scripts/base/unification/unify_loot.js:42]: 😅 I forgot to change it, sure

File: kubejs/server_scripts/base/unification/unify_loot.js (hunk 3 above)
Comment (Kanzaji) [reply kubejs/server_scripts/base/unification/unify_loot.js:24]: 👍

File: kubejs/server_scripts/base/unification/unify_loot.js (hunk 5 above)
Comment (Kanzaji) [reply kubejs/server_scripts/base/unification/unify_loot.js:46]: o.o, this I actually didn't knew about, good to know!

File: kubejs/server_scripts/base/unification/unify_loot.js (hunk 2 above)
Comment (Kanzaji) [reply kubejs/server_scripts/base/unification/unify_loot.js:11]: Main idea was to just use `typeof prefix == underfined` and it should work, however for some reason Rhino didn't like that?
But yup, `!prefix` works perfectly 😄

**Review by NielsPilgaard — APPROVED** [review]
Thanks for the fixes!
